ethereum/core/src/main/java/tech/pegasys/teku/core/BlockAttestationDataValidator.java

```diff
@@ -0,0 +1,117 @@
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_committee_count_at_slot;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_current_epoch;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_previous_epoch;
+import static tech.pegasys.teku.util.config.Constants.SLOTS_PER_EPOCH;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import javax.annotation.CheckReturnValue;
+import tech.pegasys.teku.datastructures.operations.AttestationData;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.util.config.Constants;
+
+public class BlockAttestationDataValidator {
+
+  public Optional<InvalidReason> validateAttestation(
+      final BeaconState state, final AttestationData data) {
+
+    final Supplier<Optional<InvalidReason>> justifiedCheckpointCheck;
+    if (data.getTarget().getEpoch().equals(get_current_epoch(state))) {
+      justifiedCheckpointCheck =
+          () ->
+              check(
+                  data.getSource().equals(state.getCurrent_justified_checkpoint()),
+                  InvalidReason.INCORRECT_CURRENT_JUSTIFIED_CHECKPOINT);
+    } else {
+      justifiedCheckpointCheck =
+          () ->
+              check(
+                  data.getSource().equals(state.getPrevious_justified_checkpoint()),
+                  InvalidReason.INCORRECT_PREVIOUS_JUSTIFIED_CHECKPOINT);
+    }
+
+    return firstOf(
+        () ->
+            check(
+                data.getIndex().compareTo(get_committee_count_at_slot(state, data.getSlot())) < 0,
+                InvalidReason.COMMITTEE_INDEX_TOO_HIGH),
+        () ->
+            check(
+                data.getTarget().getEpoch().equals(get_previous_epoch(state))
+                    || data.getTarget().getEpoch().equals(get_current_epoch(state)),
+                InvalidReason.NOT_FROM_CURRENT_OR_PREVIOUS_EPOCH),
+        () ->
+            check(
+                data.getTarget().getEpoch().equals(compute_epoch_at_slot(data.getSlot())),
+                InvalidReason.SLOT_NOT_IN_EPOCH),
+        () ->
+            check(
+                data.getSlot()
+                        .plus(UnsignedLong.valueOf(Constants.MIN_ATTESTATION_INCLUSION_DELAY))
+                        .compareTo(state.getSlot())
+                    <= 0,
+                InvalidReason.SUBMITTED_TOO_QUICKLY),
+        () ->
+            check(
+                state
+                        .getSlot()
+                        .compareTo(data.getSlot().plus(UnsignedLong.valueOf(SLOTS_PER_EPOCH)))
+                    <= 0,
+                InvalidReason.SUBMITTED_TOO_FAR_IN_HISTORY),
+        justifiedCheckpointCheck);
+  }
+
+  @SafeVarargs
+  private Optional<InvalidReason> firstOf(final Supplier<Optional<InvalidReason>>... checks) {
+    return Stream.of(checks)
```

Comment:
maybe you could add `Stream.parallel()` here

Reply:
The checks are quite simple so going parallel would almost certainly reduce performance because of the overhead cost. It would also be a subtle trap in the future if any of the checks depended on each other (e.g. one checks that an index is in range, one then uses that index to check a value).